### PR TITLE
Feature/402 upgrade from previous home dir version

### LINKED ISSUE #402: Copy DC configuration upon software upgrade
We have currently achieved a good level of isolation in DC's configuration by adding the version number to the home-folder path. But we also do need to consider the burden of upgrading, which is still a bit of a pain.

In the situation where we create a new home folder we should check if home folders for previous versions exist. If it does, we should take a copy of it while still overriding:

 * Example jobs
 * Example data (customers.csv)

Furthermore we should ideally assemble a new conf.xml based on the <datastore-catalog> and <reference-data-catalog> of the old conf.xml file but everything else based on the default conf.xml file.

### PR DESCRIPTION
**Pull request by tomaszguzialek**
Fixes #402 

Upgrade is performed from the newest version within the same major version. Development version of DC often starts with version set to UNKNOWN - no upgrade will be performed in such a situation.

### CODE REVIEW

File: desktop/ui/src/main/java/org/datacleaner/user/DataCleanerHomeUpgrader.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+ * Checks if there is DataCleanerHome set up for previous versions of
+ * DataCleaner and upgrades.
+ */
+public final class DataCleanerHomeUpgrader {
+    
+    private static final Logger logger = LoggerFactory.getLogger(DataCleanerHomeUpgrader.class);
+
+    private static final List<String> candidateBlacklist = Arrays.asList("log", Version.UNKNOWN_VERSION);
+
+    public boolean upgrade(FileObject target) {
```
Comment (kaspersorensen) [inline desktop/ui/src/main/java/org/datacleaner/user/DataCleanerHomeUpgrader.java]: Would be nice with a description of what (folder I assume) it is. I am not sure if it is an old folder, the new candidate or maybe their common ancestor?

**Comment by tomaszguzialek** [comment]
I added the JavaDoc to the upgrade method - is the "target" parameter clear enough now?

**Comment by kaspersorensen** [comment]
Yup, better

**Comment by kaspersorensen** [comment]
Tried running this from my local build, but am getting this... I had deleted the 4.1-SNAPSHOT folder as a preparation to see that it would be rebuilt.

```
1) Error in custom provider, org.apache.metamodel.util.ResourceException: org.apache.commons.vfs2.FileNotFoundException: Could not read from "file:///C:/Users/k
aspers/.datacleaner/4.1-SNAPSHOT/conf.xml" because it is a not a file.
  at org.datacleaner.guice.DCModuleImpl.getDataCleanerConfiguration(DCModuleImpl.java:287)
  while locating org.datacleaner.configuration.DataCleanerConfiguration

1 error
        at com.google.inject.internal.InjectorImpl$4.get(InjectorImpl.java:987)
        at com.google.inject.internal.InjectorImpl.getInstance(InjectorImpl.java:1013)
        at org.datacleaner.bootstrap.Bootstrap.runInternal(Bootstrap.java:177)
        at org.datacleaner.bootstrap.Bootstrap.run(Bootstrap.java:99)
        at org.datacleaner.Main.main(Main.java:60)
        at org.datacleaner.Main.main(Main.java:46)
Caused by: org.apache.metamodel.util.ResourceException: org.apache.commons.vfs2.FileNotFoundException: Could not read from "file:///C:/Users/kaspers/.datacleane
r/4.1-SNAPSHOT/conf.xml" because it is a not a file.
        at org.datacleaner.util.VfsResource.read(VfsResource.java:178)
        at org.datacleaner.configuration.DatastoreXmlExternalizer.<init>(DatastoreXmlExternalizer.java:77)
        at org.datacleaner.guice.DCModuleImpl$2.<init>(DCModuleImpl.java:341)
        at org.datacleaner.guice.DCModuleImpl.createDatastoreXmlExternalizer(DCModuleImpl.java:341)
        at org.datacleaner.guice.DCModuleImpl.getDataCleanerConfiguration(DCModuleImpl.java:292)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at com.google.inject.internal.ProviderMethod.get(ProviderMethod.java:104)
        at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:40)
        at com.google.inject.internal.InjectorImpl$4$1.call(InjectorImpl.java:978)
        at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1024)
        at com.google.inject.internal.InjectorImpl$4.get(InjectorImpl.java:974)
        ... 5 more
Caused by: org.apache.commons.vfs2.FileNotFoundException: Could not read from "file:///C:/Users/kaspers/.datacleaner/4.1-SNAPSHOT/conf.xml" because it is a not
a file.
        at org.apache.commons.vfs2.provider.AbstractFileObject.getInputStream(AbstractFileObject.java:1315)
        at org.apache.commons.vfs2.provider.DefaultFileContent.getInputStream(DefaultFileContent.java:396)
        at org.datacleaner.util.VfsResource.read(VfsResource.java:170)
        ... 18 more
Caused by: java.io.FileNotFoundException: C:\Users\kaspers\.datacleaner\4.1-SNAPSHOT\conf.xml (The system cannot find the file specified)
        at java.io.FileInputStream.open(Native Method)
        at java.io.FileInputStream.<init>(Unknown Source)
        at org.apache.commons.vfs2.provider.local.LocalFile.doGetInputStream(LocalFile.java:209)
        at org.apache.commons.vfs2.provider.AbstractFileObject.getInputStream(AbstractFileObject.java:1307)
        ... 20 more
```

**Pull request merged**